### ISSUE TITLE
Conversation Tree

### ISSUE THREAD

**Issue by racingcow** [collinhover/impactplusplus#111]
Hi @collinhover,

I have need of a conversation/dialogue tree in my top-down RPG game. An example might be if my PC was playing the part of [the bridgekeeper from Monty Python and the Holy Grail](http://youtu.be/pWS8Mg-JWSg)...

![conversation state transition 2](https://f.cloud.github.com/assets/1176637/1559474/54e7fd0e-4fd3-11e3-9b2b-22ca1a602e8a.png)

While I could probably achieve this by careful manipulation of the `EntityConversation.steps` array on the fly, it would be nice if there was a way for the conversation entity to handle a dialogue tree structure 'out of the box'.

If you don't already have plans for it, and don't mind me giving it a shot, I could take a look at extending `EntityConversation` to handle something like this. My thoughts on it so far are below. If you have some time, maybe you could take a look and give me some feedback/thoughts? Just let me know if there is a way that better fits with what you have done so far.
- Create a new element, perhaps called something like `UITextBubbleChoice`, that supports the user choosing one of several text items (the choices/responses that the PC would say) displayed beneath a main text heading (the last response from the NPC).
  ![uitextchoice proposed design](https://f.cloud.github.com/assets/1176637/1559464/bddfd058-4fd2-11e3-845d-569be3e4602f.png)
- When wanting to use a conversation tree, set `EntityConversation.messageContainerEntity` to `ig.UITextBubbleChoice`.
- Have each step in `EntityConversation` support another property called something like `next`. If set to an int, the conversation would move to the step represented by that index value in the `steps` array. If, on the other hand, it was set to an array of ints, the conversation would display each possible step associated with those indexes in `UITextBubbleChoice` as a possible response for the user.
- When the user chooses a step displayed in `UITextBubbleChoice`, the conversation would follow the value of `next` to display the response from the NPC.
- Support a property called something like `step.usedSteps` that would provide a list of steps that must all be visited before the step is available as a valid response (in my example step 5 should not become a valid response until all questions have been asked and answered).

If the above makes sense, this might be an example of how to set it up in Weltmeister...

``` javascript
step.1.name = 'talker'
step.1.text = 'I seek to cross the bridge.'
step.1.next = 11
step.11.name = 'player'
step.11.text = 'Stop! Who would cross the Bridge of Death must answer me these questions three, ere the other side he see.'
step.11.next = 7
step.12.name = 'talker'
step.12.text = 'Ask me the questions, bridgekeeper. I am not afraid.'
step.12.next = [2,3,4,5]
step.2.name = 'player'
step.2.text = 'What... is your name?'
step.2.next = 7
step.3.name = 'player'
step.3.text = 'What... is your quest?'
step.3.next = 8
step.4.name = 'player'
step.4.text = 'What... is your favourite colour?'
step.4.next = 9
step.5.name = 'player'
step.5.text = 'Go on. Off you go.'
step.5.usedSteps = [2,3,4]
step.5.next = 6
step.6.name = 'talker'
step.6.text = 'Oh, thank you. Thank you very much.' 
step.7.name = 'talker'
step.7.text = 'My name is Sir Lancelot of Camelot.'
step.7.next = [3,4,10,5]
step.8.name = 'talker'
step.8.text = 'To seek the Holy Grail.'
step.8.next = [2,4,10,5]
step.9.name = 'talker'
step.9.text = 'Blue.'
step.9.next = [2,3,10,5]
step.10.name = 'player'
step.10.text = 'I want to ask you something else.'
step.10.next = 12
```

**Comment by collinhover** [comment]
Neat! I don't have plans for this at the moment, and I've always found it works best to develop features within the game you need them for, so feel free to give it a shot.

It might be useful to add a few more things to each conversation step, such as `target` which is just like trigger's target property so that each step can activate triggers, and  a per step `messageContainerEntity` so that not all steps would need to be text bubbles with choices. I'd also rename `usedSteps` to `requiredSteps` for clarity.

Keep us updated =)

**Comment by racingcow** [comment]
Thanks, @collinhover. I'll do some work on it in the evenings this week and let you know when I have something to show.

**Comment by racingcow** [comment]
It took me a bit longer than I thought it would, but I have an update of sorts. Take a look at this [video of the conversation described above](http://youtu.be/6Rakx0o9D5k) inserted into the first level of the supercollider demo.

![image](https://f.cloud.github.com/assets/1176637/1643393/05fd53e2-58c5-11e3-8af7-a2035dcd1b17.png)

If this looks good, I'll try and get it into its own level in supercollider, and get a PR going soon.

Let me know what you think.

**Comment by collinhover** [comment]
Cool demo :-) looks good to me. Did you check for bugs such as layout errors when the window is resized in the middle of a conversation?

**Comment by racingcow** [comment]
No, I haven't tested that yet. Now that you mention it, I suspect there will be some more work to do around that. I'll take a look at that as soon as I can.

**Comment by collinhover** [comment]
I did make a few changes to the dev just now that fixed some resize bugs, so you may want to grab the latest when you get a chance.

**Comment by racingcow** [comment]
Thanks, @collinhover. I synced my fork up to the dev branch and re-ran my grunt script to setup supercollider to run this evening. My new conversation code definitely still has issues when resizing. I was away on business last week, but should be able to get some time this week to look into this issue. Hopefully it is something simple, and I'll have a PR soon.

**Comment by racingcow** [comment]
Okay, so after looking at this again it either isn't simple or I have missed something.

After updating [`Conversation`](https://github.com/racingcow/impactplusplus/blob/conversation-tree/lib/plusplus/entities/conversation.js) and creating [`UITextBubbleChoice`](https://github.com/racingcow/impactplusplus/blob/conversation-tree/lib/plusplus/ui/ui-text-bubble-choice.js), which derives from `UITextBubble`, I still have a resize issue.

The issue seems to be that whenever a choice (just a `UIText`) inside of the `UITextBubbleChoice` gets resized such that it takes up a different number of lines (size.y/getDrawSizeY() change), the choices aren't moved down to the new locations correctly.

If you take a look at [this video](http://youtu.be/uKTNe9h6qd4), it illustrates the issue. In this particular case, the `UITextBubbleChoice` narrows, which causes each of the choices to wrap to two lines instead of one. The height of the `UITextBubbleChoice` is correct, but for whatever reason, the choices aren't moved down to the correct locations, and instead overlap each other as if they are being left in the same positions that would be correct if they were each still a single line.

What seems strange is that when I look at the [`console.log();` messages](https://github.com/racingcow/impactplusplus/blob/conversation-tree/lib/plusplus/ui/ui-text-bubble-choice.js#L232) that I have put in there, it appears that they are being moved to the correct locations when the resize happens (at least as far as I can tell).

If you get some time, could you take a look at my code in [my fork](https://github.com/racingcow/impactplusplus/tree/conversation-tree) and see if you can see anything I am doing wrong? Perhaps I have a scaling issue going on. If you need/want to setup SUPERCOLLIDER! to reproduce the issue, I have temporarily updated **examples\supercollider\lib\game\levels\title.js** to start the game with the conversation from the example in the video.

Any help you could give to nudge me in the right direction on this would be much appreciated.

**Comment by collinhover** [comment]
Your issue may be related to #112. Let me see if I can figure out a good solution for that first and maybe it'll fix this as well.
